tidy3d/components/medium.py

```diff
@@ -548,208 +1007,142 @@ def eps_diagonal_on_grid(
+            if freq is None:
+                raise SetupError(
+                    "For a complex 'eps', 'freq' at which 'eps' is defined must be supplied",
+                )
+            eps_inf, sigma = CustomMedium.eps_complex_to_eps_sigma(eps, freq)
+            return cls(permittivity=eps_inf, conductivity=sigma, interp_method=interp_method)
+
+        # eps is ScalarFieldDataArray
+        # contradictory definition of frequency
+        freq_data = eps.coords["f"].data[0]
+        if freq is not None and not isclose(freq, freq_data):
+            raise SetupError(
+                "'freq' value is inconsistent with the coordinate 'f'"
+                "in 'eps' DataArray. It's unclear at which frequency 'eps' "
+                "is defined. Please leave 'freq=None' to use the frequency "
+                "value in the DataArray."
+            )
+        eps_inf, sigma = CustomMedium.eps_complex_to_eps_sigma(eps, freq_data)
+        eps_inf = SpatialDataArray(eps_inf.squeeze(dim="f", drop=True))
+        sigma = SpatialDataArray(sigma.squeeze(dim="f", drop=True))
+        return cls(permittivity=eps_inf, conductivity=sigma, interp_method=interp_method)
 
     @classmethod
     def from_nk(
         cls,
-        n: ScalarFieldDataArray,
-        k: Optional[ScalarFieldDataArray] = None,
+        n: Union[ScalarFieldDataArray, SpatialDataArray],
+        k: Optional[Union[ScalarFieldDataArray, SpatialDataArray]] = None,
+        freq: float = None,
         interp_method: InterpMethod = "nearest",
     ) -> CustomMedium:
         """Construct a :class:`.CustomMedium` from datasets containing n and k values.
 
         Parameters
         ----------
-        n : :class:`.ScalarFieldDataArray`
+        n : Union[:class:`.SpatialDataArray`, :class:`.ScalarFieldDataArray`]
             Real part of refractive index.
-        k : :class:`.ScalarFieldDataArray` = None
+        k : Union[:class:`.SpatialDataArray`, :class:`.ScalarFieldDataArray`], optional
             Imaginary part of refrative index.
+        freq : float, optional
+            Frequency at which ``n`` and ``k`` are defined.
         interp_method : :class:`.InterpMethod`, optional
-                Interpolation method to obtain permittivity values that are not supplied
-                at the Yee grids.
+            Interpolation method to obtain permittivity values that are not supplied
+            at the Yee grids.
 
         Returns
         -------
         :class:`.CustomMedium`
             Medium containing the spatially varying permittivity data.
         """
+        # lossless
         if k is None:
-            k = xr.zeros_like(n)
+            if isinstance(n, ScalarFieldDataArray):
+                n = SpatialDataArray(n.squeeze(dim="f", drop=True))
+            return cls(permittivity=n**2, interp_method=interp_method)
 
+        # lossy case
+        if np.logical_xor(*[isinstance(f, SpatialDataArray) for f in (n, k)]) or np.logical_xor(
```

Comment:
is it possible to replace this logic with 
```py
if not (type(n) == type(k)):
```
or even more generally 
```
if n.coords.keys() != k.coords.keys()
```
The xor logic is a bit hard to follow.

Reply:
I initially tried `type`, but pylint complains and asks me to use `isinstance`. `keys()` probably is the answer.

Reply:
sounds good